File: ios/LaneShadow/Views/Organisms/LSRouteSheet.swift
import LaneShadowTheme
import NativeTheme
import SwiftUI

/// Route sheet organism showing route details with instrument readout and weather timeline
///
/// Composes from LSBestBadge, LSScenicDotStrip, LSInstrumentReadout, LSWeatherTimeline, and LSButton.
/// Wraps content in LSBottomSheet presentation with .large detent.
public struct LSRouteSheet: View {
    @Environment(\.theme) private var theme

    public typealias Route = RouteDetails

    private let route: Route
    private let weatherTimeline: [WeatherEntry]
    private let timeRange: (String, String)
    private let onSave: () -> Void
    private let onRide: () -> Void
    private let onDismiss: () -> Void

    public init(
        route: Route,
        weatherTimeline: [WeatherEntry],
        timeRange: (String, String),
        onSave: @escaping () -> Void,
        onRide: @escaping () -> Void,
        onDismiss: @escaping () -> Void
    ) {
        self.route = route
        self.weatherTimeline = weatherTimeline
        self.timeRange = timeRange
        self.onSave = onSave
        self.onRide = onRide
        self.onDismiss = onDismiss
    }

    public var body: some View {
        VStack(spacing: 0) {
            // Scrollable content
            ScrollView {
                VStack(spacing: theme.space.md) {
                    headerSection
                    instrumentSection
                    weatherSection
                }
                .padding(.horizontal, theme.space.md)
                .padding(.top, theme.space.sm)
            }

            // Sticky action row
            actionRow
        }
        .background(LaneShadowTheme.color.surface.card)
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Route details")
        .presentationDetents([.fraction(0.9)])
        .presentationDragIndicator(.visible)
        .presentationBackground(.clear)
    }

    /// Wrapper for presenting as bottom sheet
    public func bottomSheet(
        isPresented: Binding<Bool>,
        onDismiss: @escaping () -> Void = {}
    ) -> some View {
        LSBottomSheet(
            isPresented: isPresented,
            detent: .large,
            onDismiss: onDismiss
        ) {
            self
        }
    }

    // MARK: - Header Section

    private var headerSection: some View {
        VStack(alignment: .leading, spacing: theme.space.xs) {
            // Best badge + scenic strip row
            if route.isBest {
                HStack(spacing: theme.space.sm) {
                    LSBestBadge()

                    // Scenic dot strip
                    if let scenicValue = Double(route.scenic) {
                        LSScenicDotStrip(scenicScore: scenicValue)
                    }
                }
            } else {
                // Just scenic strip for non-best routes
                if let scenicValue = Double(route.scenic) {
                    LSScenicDotStrip(scenicScore: scenicValue)
                }
            }

            // Title
            LSText(route.title, variant: .opinion.lg, color: .primary)
                .lineLimit(2)

            // Subtitle (body.sm, not body.md)
            LSText(route.subtitle, variant: .body.sm, color: .secondary)
                .lineLimit(2)
        }
    }

    // MARK: - Instrument Section

    private var instrumentSection: some View {
        LSInstrumentReadout(metrics: [
            .dist(route.distance),
            .time(route.time),
            .climb(route.climb),
            .scenic(route.scenic),
        ])
    }

    // MARK: - Weather Section

    private var weatherSection: some View {
        LSWeatherTimeline(
            entries: weatherTimeline,
            from: timeRange.0,
            to: timeRange.1
        )
    }

    // MARK: - Action Row

    private var actionRow: some View {
        GeometryReader { geometry in
            HStack(spacing: theme.space.sm) {
                // Save button (width 1/3 of available space)
                LSButton(
                    "Save",
                    variant: .outline,
                    size: .md,
                    leadingIcon: .bookmark,
                    action: onSave
                )
                .frame(width: (geometry.size.width - theme.space.sm) / 3)

                // Ride this button (width 2/3 of available space)
                LSButton(
                    "Ride this",
                    variant: .primary,
                    size: .md,
                    trailingIcon: .chevR,
                    action: onRide
                )
                .frame(width: (geometry.size.width - theme.space.sm) * 2 / 3)
            }
        }
        .frame(height: 48) // Fixed height for GeometryReader
        .padding(theme.space.md)
        .background(LaneShadowTheme.color.surface.card)
    }
}

// MARK: - Route Details Model

public struct RouteDetails: Equatable, Sendable {
    public let id: String
    public let title: String
    public let subtitle: String
    public let isBest: Bool
    public let distance: String
    public let time: String
    public let climb: String
    public let scenic: String

    public init(
        id: String,
        title: String,
        subtitle: String,
        isBest: Bool,
        distance: String,
        time: String,
        climb: String,
        scenic: String
    ) {
        self.id = id
        self.title = title
        self.subtitle = subtitle
        self.isBest = isBest
        self.distance = distance
        self.time = time
        self.climb = climb
        self.scenic = scenic
    }
}

// MARK: - Preview

#Preview("Best Route") {
    LSRouteSheet(
        route: RouteDetails(
            id: "route-1",
            title: "The Skyline Spine",
            subtitle: "via Kings Mountain Rd · Kings Mountain to Woodside",
            isBest: true,
            distance: "47",
            time: "1:22",
            climb: "3.2k",
            scenic: "4.8"
        ),
        weatherTimeline: [
            WeatherEntry(hour: "9A", condition: .clear, temp: "62°"),
            WeatherEntry(hour: "10A", condition: .clear, temp: "65°"),
            WeatherEntry(hour: "11A", condition: .clear, temp: "67°"),
            WeatherEntry(hour: "12P", condition: .wind, temp: "68°"),
            WeatherEntry(hour: "1P", condition: .wind, temp: "66°"),
            WeatherEntry(hour: "2P", condition: .clear, temp: "64°"),
        ],
        timeRange: ("9am", "3pm"),
        onSave: {},
        onRide: {},
        onDismiss: {}
    )
    .laneShadowTheme()
}

#Preview("Alt Route") {
    LSRouteSheet(
        route: RouteDetails(
            id: "route-2",
            title: "Old La Honda Road",
            subtitle: "via Page Mill Rd · Palo Alto to Woodside",
            isBest: false,
            distance: "38",
            time: "1:05",
            climb: "2.1k",
            scenic: "3.6"
        ),
        weatherTimeline: [
            WeatherEntry(hour: "9A", condition: .clear, temp: "61°"),
            WeatherEntry(hour: "10A", condition: .clear, temp: "64°"),
            WeatherEntry(hour: "11A", condition: .clear, temp: "66°"),
            WeatherEntry(hour: "12P", condition: .clear, temp: "67°"),
            WeatherEntry(hour: "1P", condition: .clear, temp: "65°"),
            WeatherEntry(hour: "2P", condition: .clear, temp: "63°"),
        ],
        timeRange: ("9am", "2pm"),
        onSave: {},
        onRide: {},
        onDismiss: {}
    )
    .laneShadowTheme()
}

#Preview("Dark Mode") {
    LSRouteSheet(
        route: RouteDetails(
            id: "route-1",
            title: "The Skyline Spine",
            subtitle: "via Kings Mountain Rd",
            isBest: true,
            distance: "47",
            time: "1:22",
            climb: "3.2k",
            scenic: "4.8"
        ),
        weatherTimeline: [],
        timeRange: ("9am", "3pm"),
        onSave: {},
        onRide: {},
        onDismiss: {}
    )
    .laneShadowTheme()
    .preferredColorScheme(.dark)
}

File: android/app/src/main/java/com/laneshadow/ui/organisms/LSRouteSheet.kt
package com.laneshadow.ui.organisms

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.SemanticsPropertyKey
import androidx.compose.ui.semantics.SemanticsPropertyReceiver
import androidx.compose.ui.semantics.semantics
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.theme.generated.LaneShadowTheme.IconName
import com.laneshadow.ui.atoms.ButtonVariant
import com.laneshadow.ui.atoms.ContentColor
import com.laneshadow.ui.atoms.LSBestBadge
import com.laneshadow.ui.atoms.LSButton
import com.laneshadow.ui.atoms.LSText
import com.laneshadow.ui.atoms.TypographyVariant
import com.laneshadow.ui.molecules.BottomSheetDetent
import com.laneshadow.ui.molecules.InstrumentMetric
import com.laneshadow.ui.molecules.LSBottomSheet
import com.laneshadow.ui.molecules.WeatherTimelineEntry

// Test tags for LSRouteSheet
const val LS_ROUTE_SHEET_TAG = "ls-route-sheet"
private const val ROUTE_SHEET_HANDLE_TAG = "route-sheet-handle"
private const val ROUTE_SHEET_BADGE_TAG = "route-sheet-badge"
private const val ROUTE_SHEET_TITLE_TAG = "route-sheet-title"
private const val ROUTE_SHEET_SUBTITLE_TAG = "route-sheet-subtitle"
private const val ROUTE_SHEET_INSTRUMENT_TAG = "ls-instrument-readout"
private const val ROUTE_SHEET_WEATHER_TAG = "ls-weather-timeline"
private const val ROUTE_SHEET_ACTIONS_TAG = "route-sheet-actions"
private const val ROUTE_SHEET_SAVE_BUTTON_TAG = "route-sheet-save-button"
private const val ROUTE_SHEET_RIDE_BUTTON_TAG = "route-sheet-ride-button"

// Semantics keys
val LSRouteSheetTitleKey = SemanticsPropertyKey<String>("LSRouteSheetTitle")
val LSRouteSheetViaKey = SemanticsPropertyKey<String>("LSRouteSheetVia")

private var SemanticsPropertyReceiver.lsRouteSheetTitle by LSRouteSheetTitleKey
private var SemanticsPropertyReceiver.lsRouteSheetVia by LSRouteSheetViaKey

// Data class for route details
data class RouteDetails(
    val id: String,
    val title: String,
    val via: String,
    val isBest: Boolean,
    val distance: String,
    val time: String,
    val climb: String,
    val scenicScore: String,
)

/**
 * LSRouteSheet - Route details bottom sheet organism
 *
 * Composed from LSBottomSheet with drag handle, optional LSBestBadge,
 * opinion-serif title, instrument readout, weather timeline, and sticky action row.
 *
 * @param route Route details including title, via, isBest flag, and metrics
 * @param weatherTimeline List of weather timeline entries for the route
 * @param timeRange Time range for the weather timeline header as Pair<from, to>
 * @param onSave Callback when Save button is tapped
 * @param onRide Callback when Ride this button is tapped
 * @param onDismiss Callback when sheet is dismissed (drag-down or backdrop tap)
 * @param modifier Modifier for the root composable
 */
@Composable
fun LSRouteSheet(
    route: RouteDetails,
    weatherTimeline: List<WeatherTimelineEntry>,
    timeRange: Pair<String, String>,
    onSave: () -> Unit,
    onRide: () -> Unit,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    LSBottomSheet(
        detent = BottomSheetDetent.Large,
        onDismiss = onDismiss,
        modifier = modifier.testTag(LS_ROUTE_SHEET_TAG),
    ) {
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(theme.space.md),
        ) {
            // Header row: best badge (optional) + title + subtitle
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag(ROUTE_SHEET_HEADER_TAG),
                verticalArrangement = Arrangement.spacedBy(theme.space.xs),
            ) {
                // Best badge (only when route.isBest == true)
                if (route.isBest) {
                    Box(
                        modifier = Modifier.testTag(ROUTE_SHEET_BADGE_TAG),
                        contentAlignment = Alignment.CenterStart,
                    ) {
                        LSBestBadge()
                    }
                }

                // Title (opinion.lg)
                LSText(
                    text = route.title,
                    variant = TypographyVariant.Opinion.Lg,
                    color = ContentColor.Primary,
                    modifier = Modifier
                        .testTag(ROUTE_SHEET_TITLE_TAG)
                        .semantics { lsRouteSheetTitle = route.title },
                )

                // Via subtitle (ui.body.md, textMuted)
                LSText(
                    text = route.via,
                    variant = TypographyVariant.Ui.Body.Md,
                    color = ContentColor.Subtle,
                    modifier = Modifier
                        .testTag(ROUTE_SHEET_SUBTITLE_TAG)
                        .semantics { lsRouteSheetVia = route.via },
                )
            }

            // Instrument readout (4 metrics: dist, time, climb, scenic)
            com.laneshadow.ui.molecules.LSInstrumentReadout(
                metrics = listOf(
                    InstrumentMetric(label = "Dist", value = route.distance),
                    InstrumentMetric(label = "Time", value = route.time),
                    InstrumentMetric(label = "Climb", value = route.climb),
                    InstrumentMetric(label = "Scenic", value = route.scenicScore, isAccent = route.isBest),
                ),
                modifier = Modifier.testTag(ROUTE_SHEET_INSTRUMENT_TAG),
            )

            // Weather timeline
            com.laneshadow.ui.molecules.LSWeatherTimeline(
                entries = weatherTimeline,
                from = timeRange.first,
                to = timeRange.second,
                modifier = Modifier.testTag(ROUTE_SHEET_WEATHER_TAG),
            )

            // Sticky action row: Save (outline, flex 1) + Ride this (primary, flex 2)
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag(ROUTE_SHEET_ACTIONS_TAG),
                horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
            ) {
                // Save button (outline, flex 1)
                LSButton(
                    label = "Save",
                    variant = ButtonVariant.Outline,
                    leadingIcon = IconName.Bookmark,
                    onClick = onSave,
                    modifier = Modifier
                        .weight(1f)
                        .testTag(ROUTE_SHEET_SAVE_BUTTON_TAG),
                )

                // Ride this button (primary, flex 2)
                LSButton(
                    label = "Ride this",
                    variant = ButtonVariant.Primary,
                    trailingIcon = IconName.ChevR,
                    onClick = onRide,
                    modifier = Modifier
                        .weight(2f)
                        .testTag(ROUTE_SHEET_RIDE_BUTTON_TAG),
                )
            }
        }
    }
}

private const val ROUTE_SHEET_HEADER_TAG = "route-sheet-header"
